### PR TITLE
refactor(cc): split up `CongestionEvent` enum

### PR DESCRIPTION
**Pull request by omansfeld**
- into mutually exclusive `CongestionTrigger` enum for control flow and `CongestionEventStats` struct that includes orthogonal categories for stat recording

As discussed in https://github.com/mozilla/neqo/pull/3527#issuecomment-4205870835 (see below), but only the refactor without the new metric that exists in that PR.

> [...] How about the following middle ground:
> 
> Keeping the enum change, so we have the mutually exclusive `Loss` and `Ecn` used in control flow:
> 
> ```rust
> pub enum CongestionTrigger { Loss, Ecn }
> ```
> 
> And then simplify the stat side:
> 
> ```rust
> pub struct CongestionEventStats {
>     /// Congestion events triggered by packet loss.
>     pub loss: usize,
>     /// Congestion events triggered by ECN-CE marks.
>     pub ecn: usize,
>     /// Events later found spurious (subset of loss+ecn total).
>     pub spurious: usize,
>     /// Events while underutilized (subset of loss+ecn total).
>     pub underutilized: usize,
>   }
> ```
> 
> This way we wouldn't have those `cc_stats.congestion_events.by_trigger[CongestionTrigger::Loss]` calls anymore (which is I think part of your concern in added complexity?) but still have the type-safety in actual control flow with `CongestionTrigger`.
> 
> Alternatively I could just add `Underutilized` to the existing `CongestionEvent` enum, but then the problems (see below) I outlined remain. I'm not a big fan of those.
> 
> > I think we shouldn't mix triggers (ecn, loss) with categories for stats (triggered by ecn/loss, spurious, underutilized) in a single enum. Right no we e.g. use CongestionEvent for on_congestion_event calls where we only expect Loss or Ecn to be passed in. I think it would be good to actually codify that expectation by having an extra enum for CongestionTriggers, especially so if we add more categories like we do in this patch.
> 
> > I also think it is slightly confusing to have a congestion_events EnumMap where the total sum of all entries doesn't equal the total sum of all congestion events (which it doesn't because a congestion event can be both triggered by loss, spurious and triggered while underutilized). But maybe that's a problem on my intuition about enums.

This would obviously need changes to glue code on the Firefox side since it restructures stat fields consumed by Gecko.

### CODE REVIEW

File: neqo-transport/src/stats.rs [hunk 1]
```diff
@@ -142,16 +142,26 @@ pub enum SlowStartExitReason {
     Heuristic,
 }
 
+/// Congestion event counters.
+///
+/// `loss` and `ecn` are mutually exclusive triggers (their sum equals the total number of
+/// congestion events). `spurious` is an orthogonal category that can overlap with any trigger.
+#[derive(Default, Clone, PartialEq, Eq)]
```
[the remark this answers is not in the document]
Comment (omansfeld) [reply neqo-transport/src/stats.rs:150]: Other way around: `SlowStartExitReason` is the only such field that has a `Debug` derive (because it needs one).

File: neqo-transport/src/cc/classic_cc.rs [hunk 2]
```diff
@@ -881,7 +881,10 @@ where
             self.current.ssthresh
         );
 
-        cc_stats.congestion_events[congestion_event] += 1;
+        match congestion_trigger {
+            Loss => cc_stats.congestion_events.loss += 1,
+            Ecn => cc_stats.congestion_events.ecn += 1,
+        }
```
[the remark this answers is not in the document]
Comment (omansfeld) [reply neqo-transport/src/cc/classic_cc.rs:886]: Clippy would actually complain about this semicolon.

File: neqo-transport/src/stats.rs [hunk 3]
```diff
@@ -142,16 +142,26 @@ pub enum SlowStartExitReason {
     Heuristic,
 }
 
+/// Congestion event counters.
+///
+/// `loss` and `ecn` are mutually exclusive triggers (their sum equals the total number of
+/// congestion events). `spurious` is an orthogonal category that can overlap with any trigger.
```
[the remark this answers is not in the document]
Comment (omansfeld) [reply neqo-transport/src/stats.rs]: It could overlap in theory, as in the setup of `CongestionEventStats` is meant in a way that orthogonal fields such as `spurious` can overlap with the mutually exclusive triggers and don't count into total congestion event sum.

**Comment by omansfeld** [comment]
@larseggert the new `github-actions` review is Claude? Very nice that it seems to work now!

File: neqo-transport/src/cc/classic_cc.rs [hunk 4]
```diff
@@ -860,7 +859,7 @@ where
             return false;
         }
 
-        if congestion_event != CongestionEvent::Ecn {
+        if congestion_trigger != Ecn {
```
[the remark this answers is not in the document]
Comment (omansfeld) [reply neqo-transport/src/cc/classic_cc.rs:862]: Imo "can't have spurious congestion for ecn" is also a fair way to read this and how it was originally intended.

**Comment by larseggert** [comment]
> @larseggert the new `github-actions` review is Claude? Very nice that it seems to work now!

Yep. You can edit the prompt (via a PR) to tweak what it's doing.

**Review by larseggert — APPROVED** [review]
